_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,181 @@
+date: 2019-08-08
+author: kyle_liberti
+---
+
+Cruise Control is a higly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Controls’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working on a Stimzi deployed Kafka cluster.
+
+# Prequisite Outline
+
+Like any proof of concept, there are a few actions to consider when putting it together. For getting Cruise Control to work with Strimzi these actions are: 
+
+- Aligning Kafka versions
+- Gathering Kafka Metrics
+- Accessing Zookeeper endpoint
+- Containerizing and deploying Cruise Control
+
+# Aligning Kafka versions
+
+Its worth noting that the latest Kafka version which Cruise Control supports at this time is 2.0.1. The latest version of Strimzi which supports Kafka version 2.0.1 is 0.11.4. Although its possible to run Cruise Control alongside with later versions of Kafka, 2.0.1+, it is not supported. Keep this in mind when deciding which version of Strimzi to deploy. This post will stick to using Strimzi 0.11.4, but feel free to experiment with more updated versions.  
+
+# Gathering Metrics
+
+In order to estimate partition activity and resources, Cruise Control needs to a steady source of raw metrics from Kafka. Cruise Control also needs a method of processing and a place of storing these raw metrics as well. Luckily, Cruise Control ships with an implemenation of a Kafka Metrics Reporter which takes care of these things. Making this metric sampler available to Strimzi Kafka brokers requires creating a new Kafka image and placing the metrics reporter in the `/opt/kafka/libs/` directory.
+
+First create a Dockerfile like this:
+```Dockerfile
+FROM strimzi/kafka:0.11.4-kafka-2.0.1
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"
+
+USER root
+
+RUN yum -y install git java-1.8.0-openjdk-devel \
+    && yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+RUN cd cruise-control \
+    && ./gradlew jar :cruise-control-metric-reporter:jar \
+    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
+```
+
+Then build the image:
+
+```bash
+docker build Dockerfile -t <kafka-image-name>
+```
+
+Create a basic Kafka resource, `kafka.yaml`, then specify the Cruise Control metrics reporter in the Kafka `metrics.reporters` config:
+```yaml
+apiVersion: kafka.strimzi.io/v0alpha1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    image: <kafka-image-name>
+    version: 1.0.1
+    replicas: 2
+    listeners:
+      plain: {}
+      tls: {}
+    config:
+      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
+      offsets.topic.replication.factor: 0
```

Comment:
I'm quite sure that the `0` here is not standard. Does it even work with `0`?

Reply:
You are right, fixed it to one. Surprisingly it worked, it just defaulted to 1.